Jwt mcl

matclab opened the pull request:
Some improvement to jwt in web2py:
- Allows to encode date and datetime with json (needed when extra_fields of type date or datetime used for auth_user table)
- Allows jwt token to be produced with the logged in user …

This allows for jwt authentication without transmitting user and
password in clear. A use case is:

 - user first logs in,
 - he gets the token online over https,
 - he then issues API requests with the token as a HTTP header.

matclab commented:
Deserialization of date and datetime is not taken into account.
We should probably use gluon.serialize somewhere…

mdipierro commented on gluon/tools.py:
```diff
@@ -1393,16 +1400,16 @@ def api_auth():
-                    401, u'Not Authorized', 
+                    401, u'Not Authorized',
                     **{'WWW-Authenticate': u'JWT realm="%s"' % self.realm})
-        else:
```
why is this removed? shouldn't we handle this case?

matclab replied:
It seems I've messed the PR. There is some indentation mess.
The idea was to keep only the 'Not Authorized' redirection…
I'll have a second take to this PR… after a good night.

matclab commented:
@mdipierro what is the *web2py* way to cast a string back to a field type ? 
I'd like to do the following in `inject_token`:
```python
     self.auth.user = Storage(tokend['user'])
     for field in self.auth.settings.extra_fields[self.auth.settings.table_user_name]:
         self.auth.user[field.name] = cast(field.type, self.auth.user[field.name])
```
But I do not know how to do the `cast`.
Is this transformation already done sowewhere in web2py ?

mdipierro commented:
You cannot really cast that way because the conversion may not be possible. I would not do the casting.

matclab commented:
@mdipierro I've taken into account your comment.

The new commits, only make use of gluon.serializers in order to take into account more types when serializing to json, and allow to generate the token in the browser when connected. 
I've also corrected a typo and added a comment about using before_authorization callback for extra fields…

niphlod commented:
thanks for the PR. I needed to make further modifications to it so I raised https://github.com/web2py/web2py/pull/1149 that contains this and another fix.

matclab commented:
Thanks !
